### PR TITLE
Fix #3041 - Random crashes when you spawn/switch/delete actor

### LINKED ISSUE #3041: Random crashes when switching/spawning/deleting actors
Issue ticket for the random crashes that occur when spawning/deleting/switching actors, originally brought up by @tritonas00 in #development on Discord. Related PR: #3042 

I'm able to consistently reproduce this by loading [Windmill Mountain](https://forum.rigsofrods.org/resources/windmill-mountain.500/) and switching to the first "Wind Engine Tower" on the vehicle list, the game will crash a few seconds after. Oddly enough, attempting to load that terrain with a Debug build crashes with: 
devenv_2023-05-06_12-57-45
(still not sure of the correct way to share VS back traces)
Release build back trace after switching to the wind tower actor:
devenv_2023-05-06_13-07-02

### PR DESCRIPTION
**Pull request by ohlidalp**
See #3041
I fixed one potential crash due to dangling GfxActor pointer (hook holding pointer to DISPOSED actor which doesn't have GfxActor anymore). 

UPDATE: This evolved into signifficant cleanup of the Actor internals. The motivation was to use safer memory handling with added debug checks, but it will also help future extensions.  The original problem turned out to be using non-threadsafe `RefCountingObject<>`s from physics worker threads - see conclusion: https://github.com/RigsOfRods/rigs-of-rods/pull/3042#issuecomment-1557066997

UPDATE2 (2023-05-29): I surrendered and threw a mutex in `RefCountingObject`. Not proud of this at all, but I apparently used it wrong in too many places to look for it now. Technical debt ahoy!

UPDATE3 (2023-06-18): Closing in favor of #3057 (fixes the crashes) and #3059 (refactors the Actor memory management).

### CODE REVIEW

**Comment by tritonas00** [comment]
```
Thread 1 "RoR" received signal SIGABRT, Aborted.
0x00007ffff649f26c in ?? () from /usr/lib/libc.so.6
(gdb) bt
#0  0x00007ffff649f26c in ?? () from /usr/lib/libc.so.6
#1  0x00007ffff644fa08 in raise () from /usr/lib/libc.so.6
#2  0x00007ffff6438538 in abort () from /usr/lib/libc.so.6
#3  0x00007ffff64392db in ?? () from /usr/lib/libc.so.6
#4  0x00007ffff64a91b7 in ?? () from /usr/lib/libc.so.6
#5  0x00007ffff64ab484 in ?? () from /usr/lib/libc.so.6
#6  0x00007ffff64adcb3 in free () from /usr/lib/libc.so.6
#7  0x0000555555865f9f in RefCountingObject<RoR::SoundScriptInstance>::Release (this=<optimized out>)
    at /home/babis/Downloads/ror-dependencies/rigs-of-rods/source/main/utils/memory/RefCountingObject.h:46
#8  RefCountingObject<RoR::SoundScriptInstance>::Release (this=<optimized out>)
    at /home/babis/Downloads/ror-dependencies/rigs-of-rods/source/main/utils/memory/RefCountingObject.h:40
#9  RefCountingObjectPtr<RoR::SoundScriptInstance>::ReleaseHandle (this=0x5555618b8f48)
    at /home/babis/Downloads/ror-dependencies/rigs-of-rods/source/main/utils/memory/RefCountingObjectPtr.h:211
#10 RefCountingObjectPtr<RoR::SoundScriptInstance>::Set (ref=0x0, this=0x5555618b8f48)
    at /home/babis/Downloads/ror-dependencies/rigs-of-rods/source/main/utils/memory/RefCountingObjectPtr.h:241
#11 RefCountingObjectPtr<RoR::SoundScriptInstance>::operator= (other=..., this=0x5555618b8f48)
    at /home/babis/Downloads/ror-dependencies/rigs-of-rods/source/main/utils/memory/RefCountingObjectPtr.h:231
#12 RoR::Actor::dispose (this=this@entry=0x5555618a5c00) at /home/babis/Downloads/ror-dependencies/rigs-of-rods/source/main/physics/Actor.cpp:117
#13 0x000055555587ada0 in RoR::ActorManager::DeleteActorInternal (this=this@entry=0x555555c72730 <RoR::App::g_game_context+80>, actor=...)
    at /home/babis/Downloads/ror-dependencies/rigs-of-rods/source/main/utils/memory/RefCountingObjectPtr.h:51
#14 0x000055555562c1e9 in RoR::GameContext::DeleteActor (this=<optimized out>, actor=...)
    at /home/babis/Downloads/ror-dependencies/rigs-of-rods/source/main/GameContext.cpp:429
#15 0x00005555555f46a0 in main (argc=<optimized out>, argv=<optimized out>)
    at /home/babis/Downloads/ror-dependencies/rigs-of-rods/source/main/main.cpp:688
    ```

**Comment by only-a-ptr** [comment]
I've piled up a series of "pointer -> index" refactoring commits here. I needed the Actor's node/beam buffers to become `std::vector<>`s to get the advantage of bounds checking and also get rid of `memset()`ing. I wanted to do this anyway for cleaner code and to make Actors editable. 

Interestingly, I can't reproduce the "start 3 AI agoras, then stop, after a few tries -> boom" scenario anymore.

File: source/main/physics/ActorManager.cpp [hunk 1]
```diff
@@ -679,11 +679,22 @@ void ActorManager::ForwardCommands(ActorPtr source_actor)
                 }
                 if (source_actor->ar_toggle_ties)
                 {
-                    actor->tieToggle();
+                    //actor->tieToggle();
+                    ActorLinkingRequest* rq = new ActorLinkingRequest();
+                    rq->alr_type = ActorLinkingRequestType::TIE_ACTION;
+                    rq->alr_actor_instance_id = source_actor->ar_instance_id;
```
Comment (po185021) [inline source/main/physics/ActorManager.cpp:685]: oops... bad actor

File: source/main/physics/ActorManager.cpp [hunk 2]
```diff
@@ -679,11 +679,22 @@ void ActorManager::ForwardCommands(ActorPtr source_actor)
...
+                    App::GetGameContext()->PushMessage(Message(MSG_SIM_ACTOR_LINKING_REQUESTED, rq));
+
                 }
                 if (source_actor->ar_toggle_ropes)
                 {
-                    actor->ropeToggle(-1);
+                    //actor->ropeToggle(-1);
+                    ActorLinkingRequest* rq = new ActorLinkingRequest();
+                    rq->alr_type = ActorLinkingRequestType::ROPE_ACTION;
+                    rq->alr_actor_instance_id = source_actor->ar_instance_id;
```
Comment (po185021) [inline source/main/physics/ActorManager.cpp:695]: ditto

**Comment by CuriousMike56** [comment]
Spamming start/stop on AI gives me:
image

File: source/main/physics/ActorManager.cpp (hunk 1 above)
Comment (only-a-ptr) [reply source/main/physics/ActorManager.cpp]: fixed

File: source/main/physics/ActorManager.cpp (hunk 2 above)
Comment (only-a-ptr) [reply source/main/physics/ActorManager.cpp]: fixed

**Comment by only-a-ptr** [comment]
It should be ready.

Turns out I was really careless when I made `class Actor` a `RefCountingObject<>` - sure, the refcounting was broken before and this helper works flawlessly, except you have to remember not to use it from threads, and I did.

**Comment by CuriousMike56** [comment]
Windmill Mountain crashes on final "Loading actors" stage (Release)
image
Debug result:
image
image

Debug: F1 track, "Grand prix long" preset, 2 agoras, spamming start/stop eventually gives 
image
Also got this nasty debug error the first time, no back trace:
image

**Comment by only-a-ptr** [comment]
Closing in favor of https://github.com/RigsOfRods/rigs-of-rods/pull/3057 (fixes the crashes) and https://github.com/RigsOfRods/rigs-of-rods/pull/3059 (refactors the Actor memory management).

**Pull request closed without merging**